cpp/src/arrow/table_builder.cc

```diff
@@ -62,7 +62,21 @@ Status RecordBatchBuilder::Flush(bool reset_builders,
     }
     length = fields[i]->length();
   }
-  *batch = RecordBatch::Make(schema_, length, std::move(fields));
+
+  // For certain types like dictionaries, types may not be fully
+  // determined before we have flushed. Make sure that the RecordBatch
+  // gets the correct types in schema.
+  // See: #ARROW-9969
+  std::shared_ptr<Schema> schema(schema_);
+  std::vector<std::shared_ptr<Field>> schema_fields(schema->fields());
```

Comment:
No need to copy the vector around. Just:
```c++
const auto& schema_fields = schema->fields();
```

Reply:
I've changed the logic so that the vector might be modified, and now it need to be copied.